File: pkg/rtweekend/camera.go
package rtweekend

import (
	"math"
)

// Camera object
type Camera struct {
	origin          Point3
	lowerLeftCorner Point3
	horizontal      Vec3
	vertical        Vec3
	u, v, w         Vec3
	lensRadius      float64
	time0, time1    float64
}

/* Returns pointer to a Camera object
   Params
       lookFrom     Point3      Coordinate where the camera is
       lookAt       Point3      Coordinate where the scene is
       vup          Vec3        Direction for view up
       vfov         float64     Vertical field-of-view in degrees
       aspectRatio  float64     Aspect ratio
       aperture     float64     Aperture
       focusDist    float64     Focus distance for camera
       time         []float64   Start and end point of time (both defaults to 0)
*/
func NewCamera(
	lookFrom, lookAt Point3,
	vup Vec3,
	vfov, aspectRatio, aperture, focusDist float64,
	time ...float64,
) *Camera {
	camera := new(Camera)

	if len(time) == 2 {
		camera.time0 = time[0]
		camera.time1 = time[1]
	} else if len(time) == 1 {
		camera.time0 = time[0]
		camera.time1 = 0
	} else {
		camera.time0 = 0
		camera.time1 = 0
	}
	theta := DegreesToRadians(vfov)
	h := math.Tan(theta / 2)
	viewportHeight := 2.0 * h
	viewportWidth := aspectRatio * viewportHeight

	camera.w = *UnitVector(lookFrom.Substract(&lookAt))
	camera.u = *UnitVector(Cross(&vup, &camera.w))
	camera.v = *Cross(&camera.w, &camera.u)

	camera.origin = lookFrom
	camera.horizontal = *camera.u.Multiply(focusDist * viewportWidth)
	camera.vertical = *camera.v.Multiply(focusDist * viewportHeight)
	camera.lowerLeftCorner = *camera.origin.Substract(
		camera.horizontal.Divide(2),
	).Substract(
		camera.vertical.Divide(2),
	).Substract(camera.w.Multiply(focusDist))
	camera.lensRadius = aperture / 2

	return camera
}

/* Computes the rays from camera viewpoint and angle
   Param
       s float64   Horizontal ray length
       t float64   Vertical ray length
*/
func (c *Camera) GetRay(s, t float64) *Ray {
	rd := RandomInUnitDisk().Multiply(c.lensRadius)
	offset := c.u.Multiply(rd.X()).Add(c.v.Multiply(rd.Y()))

	return NewRay(
		*c.origin.Add(offset),
		*c.lowerLeftCorner.Add(
			c.horizontal.Multiply(s),
		).Add(
			c.vertical.Multiply(t),
		).Substract(
			&c.origin,
		).Substract(offset),
		RandomFloat64InRange(c.time0, c.time1),
	)
}

File: src/lib/camera.rs
use crate::{cross, degrees_to_radians, random_in_unit_disk, unit_vector};
use crate::{Point3, Ray, Vec3};

pub struct Camera {
    origin: Point3,
    lower_left_corner: Point3,
    horizontal: Vec3,
    vertical: Vec3,
    u: Vec3,
    v: Vec3,
    w: Vec3,
    lens_radius: f64,
}

impl Camera {
    pub fn new(
        look_from: Point3,
        look_at: Point3,
        vup: Vec3,
        vfov: f64, // vertical field of view
        aspect_ratio: f64,
        aperture: f64,
        focus_distance: f64,
    ) -> Self {
        let theta = degrees_to_radians(vfov);
        let h = (theta / 2.0).tan();
        let viewport_height = 2.0 * h;
        let viewport_width = aspect_ratio * viewport_height;

        let w = unit_vector(&(look_from - look_at));
        let u = unit_vector(&cross(&vup, &w));
        let v = cross(&w, &u);

        let origin = look_from;
        let horizontal = focus_distance * viewport_width * u;
        let vertical = focus_distance * viewport_height * v;
        let lower_left_corner = origin - horizontal / 2.0 - vertical / 2.0 - focus_distance * w;
        let lens_radius = aperture / 2.0;

        Self {
            origin,
            lower_left_corner,
            horizontal,
            vertical,
            u,
            v,
            w,
            lens_radius,
        }
    }

    pub fn get_ray(&self, u: f64, v: f64) -> Ray {
        let rd = self.lens_radius * random_in_unit_disk();
        let offset = self.u * rd.x() + self.v * rd.y();

        Ray::new(
            self.origin + offset,
            self.lower_left_corner + u * self.horizontal + v * self.vertical - self.origin - offset,
        )
    }
}
